File: ipd.go
package main

import (
	"fmt"
)

type Decision uint

const (
	Defect = iota
	Cooperate
)

type Prisoner struct {
	Name       string
	Decisions  []Decision
	Scores     []uint
	TotalScore uint64

	Strategy
}

type Prisoners [2]*Prisoner

type Strategy interface {
	Move(opponent *Prisoner) Decision
}

type Trial [2]Decision

var ScoreMatrix = map[Trial]uint{
	Trial{Cooperate, Defect}:    0,
	Trial{Defect, Defect}:       1,
	Trial{Cooperate, Cooperate}: 3,
	Trial{Defect, Cooperate}:    5,
}

type AlwaysDefect struct{}
type AlwaysCooperate struct{}

func (s AlwaysDefect) Move(opponent *Prisoner) Decision {
	return Defect
}

func (s AlwaysCooperate) Move(opponent *Prisoner) Decision {
	return Cooperate
}

func runGame(prisoners *Prisoners, times int) {
	toBoth := func(ps *Prisoners, fn func(mover, opp *Prisoner)) {
		fn(ps[0], ps[1])
		fn(ps[1], ps[0])
	}

	for i := 0; i < times; i++ {
		toBoth(prisoners, func(mover, opp *Prisoner) {
			mover.Decisions = append(mover.Decisions, mover.Move(opp))
		})
	}

	for i := 0; i < len(prisoners[0].Decisions); i++ {
		toBoth(prisoners, func(mover, opp *Prisoner) {
			score := ScoreMatrix[Trial{mover.Decisions[i], opp.Decisions[i]}]

			mover.Scores = append(mover.Scores, score)
			mover.TotalScore += uint64(score)
		})
	}
}

func main() {
	prisoners := &Prisoners{
		&Prisoner{Name: "a", Strategy: new(AlwaysCooperate)},
		&Prisoner{Name: "b", Strategy: new(AlwaysDefect)},
	}

	runGame(prisoners, 10)

	fmt.Println("Prisoner A (AlwaysCooperate):", prisoners[0].TotalScore)
	fmt.Println("Prisoner B (AlwaysDefect)   :", prisoners[1].TotalScore)
}

File: ipd.swift
enum DecisionType {
	case Defect
	case Cooperate
}

protocol Strategy {
	func move(opponent: Prisoner) -> DecisionType
}

struct AlwaysDefect : Strategy {
	func move(opponent: Prisoner) -> DecisionType {
		return DecisionType.Defect
	}
}

struct AlwaysCooperate : Strategy {
	func move(opponent: Prisoner) -> DecisionType {
		return DecisionType.Cooperate
	}
}

struct Trial {
	subscript(mover: DecisionType, opp: DecisionType) -> Int {
		get {
			switch (mover, opp) {
			case (.Cooperate, .Defect):
				return 0
			case (.Defect, .Defect):
				return 1
			case (.Cooperate, .Cooperate):
				return 3
			case (.Defect, .Cooperate):
				return 5
			}
		}
	}
}

class Prisoner : Strategy {
	let name: String
	let strategy: Strategy
	var scores: [Int] = []
	var decisions: [DecisionType] = []
	var totalScore: Int = 0
	
	init(_ name: String, _ strategy: Strategy) {
		self.name = name
		self.strategy = strategy
	}
	
	func move(mover: Prisoner) -> DecisionType {
		return strategy.move(mover)
	}
}

func runGame(ps: [Prisoner], times: Int) -> () -> (String, String) {
	return { results in
		func swap(ps: [Prisoner], fn: (mover: Prisoner, opp: Prisoner) -> ()) {
			fn(mover: ps[0], opp: ps[1])
			fn(mover: ps[1], opp: ps[0])
		}
		
		for i in 0..<times {
			swap(ps) { (mover: Prisoner, opp: Prisoner) in
				mover.decisions.append(mover.move(opp))
			}
			
			swap(ps) { (mover: Prisoner, opp: Prisoner) in
				let score = Trial()[mover.decisions[i], opp.decisions[i]]
				mover.scores.append(score)
				mover.totalScore += score
			}
		}
		
		return ("Prisoner A (AlwaysCooperate): \(prisoners[0].totalScore)", "Prisoner B (AlwaysDefect): \(prisoners[1].totalScore)")
	}
}

let prisoners: [Prisoner] = [Prisoner("a", AlwaysCooperate()), Prisoner("b", AlwaysDefect())]
let (PrisonerA, PrisonerB) = runGame(prisoners, 10)()
println(PrisonerA)
println(PrisonerB)
